controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -932,3 +932,224 @@ func (r *Reconciler) attachmentsToDelete(
+	// want this behavior. It can be nice though to show detaching volumes. It would be nice if
+	// the Volume status has a reference to the CnsNodeVmAttachment.
+	// Once its attachment is deleted, the volume status will be removed.
+	for _, volume := range ctx.VM.Status.Volumes {
+		if attachment, ok := uuidAttachments[volume.DiskUUID]; ok {
+			volName := volume.Name
+			if !strings.HasSuffix(volName, volumeNameDetachSuffix) {
+				volName += volumeNameDetachSuffix
+			}
+			volumeStatuses = append(volumeStatuses, legacyAttachmentToVolumeStatus(volName, attachment))
+		}
+	}
+
+	return volumeStatuses
+}
+
+// getVolumeStatusWithDetachingVolumeFromBatchAttachment gets any disks that are
+// in detaching state in batchAttachment.status to the volumeStatuses.
+// These volumes that are being detached all has same suffix ':detaching'.
+func getVolumeStatusWithDetachingVolumeFromBatchAttachment(
+	existingAttachVolStatus map[string]cnsv1alpha1.VolumeStatus,
+	volumeStatuses *[]vmopv1.VirtualMachineVolumeStatus) {
+
+	for _, volStatus := range existingAttachVolStatus {
+		volName := volStatus.Name
+		if strings.HasSuffix(volName, volumeNameDetachSuffix) {
+			*volumeStatuses = append(*volumeStatuses,
+				attachmentStatusToVolumeStatus(volName, volStatus),
+			)
+		}
+	}
+}
+
+func attachmentStatusToVolumeStatus(
+	volName string,
+	volStatus cnsv1alpha1.VolumeStatus) vmopv1.VirtualMachineVolumeStatus {
+
+	return vmopv1.VirtualMachineVolumeStatus{
+		Name:     volName,
+		Attached: volStatus.PersistentVolumeClaim.Attached,
+		DiskUUID: volStatus.PersistentVolumeClaim.DiskUUID,
+		Error:    pkgutil.SanitizeCNSErrorMessage(volStatus.PersistentVolumeClaim.Error),
+		Type:     vmopv1.VolumeTypeManaged,
+	}
+}
+
+func legacyAttachmentToVolumeStatus(
+	volumeName string,
+	attachment cnsv1alpha1.CnsNodeVmAttachment) vmopv1.VirtualMachineVolumeStatus {
+
+	return vmopv1.VirtualMachineVolumeStatus{
+		Name:     volumeName, // Name of the volume as in the Spec
+		Attached: attachment.Status.Attached,
+		DiskUUID: attachment.Status.AttachmentMetadata[cnsv1alpha1.AttributeFirstClassDiskUUID],
+		Error:    pkgutil.SanitizeCNSErrorMessage(attachment.Status.Error),
+		Type:     vmopv1.VolumeTypeManaged,
+	}
+}
+
+// processLegacyAttachmentsAndReturnVolumeStatus processes the legacy attachments and returns the
+// volume statuses that are tracked by the legacy attachments.
+func (r *Reconciler) processLegacyAttachmentsAndReturnVolumeStatus(
+	ctx *pkgctx.VolumeContext,
+	legacyAttachments map[string]cnsv1alpha1.CnsNodeVmAttachment,
+	orphanedAttachments []cnsv1alpha1.CnsNodeVmAttachment,
+	existingVMManagedVolStatus map[string]vmopv1.VirtualMachineVolumeStatus,
+	vmVolumeSpecsForLegacy []vmopv1.VirtualMachineVolume,
+) []vmopv1.VirtualMachineVolumeStatus {
+
+	volumeStatuses := []vmopv1.VirtualMachineVolumeStatus{}
+
+	// Make a copy here since we might need to skip adding some of the orphaned
```

Comment:
"Make a copy" is somewhat misleading here, as it is still the same underlying slice. 
slices.DeleteFunc() below does return a new slice, tho, so the functionality is correct. 

The concern is that the code in the future might mutate elements directly.

Reply:
Good catch. Let me make it a map as you commented below. Forgot about actually `copy()` the array